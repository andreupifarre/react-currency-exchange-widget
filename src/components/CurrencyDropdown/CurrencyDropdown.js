import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import './CurrencyDropdown.scss';

const CurrencyDropdown = ({ currency, currencies, onCurrencyChange }) => {
  const [shown, setShown] = useState(false);

  const handleClick = (item) => {
    setShown(false);
    onCurrencyChange(item);
  };

  return (
    <div className="currency-dropdown">
      <span
        className="currency-dropdown__button"
        onClick={() => setShown(!shown)}
      >
        {currency}
      </span>
      <ul
        className={classNames('currency-dropdown__list', {
          'currency-dropdown__list--shown': shown,
        })}
      >
        {currencies.map((c, i) => (
          <li
            key={c.code}
            className={classNames('currency-dropdown__item', {
              'currency-dropdown__item--active': currency === c.code,
            })}
            onClick={() => handleClick(c.code)}
          >
            {c.code} â€¢ {c.sign}
            {c.balance}
          </li>
        ))}
      </ul>
    </div>
  );
};

export default CurrencyDropdown;

CurrencyDropdown.defaultProps = {
  currencies: [],
  currency: '',
  onCurrencyChange: () => {},
};

CurrencyDropdown.propTypes = {
  currencies: PropTypes.array.isRequired,
  currency: PropTypes.string.isRequired,
  onCurrencyChange: PropTypes.func.isRequired,
};
