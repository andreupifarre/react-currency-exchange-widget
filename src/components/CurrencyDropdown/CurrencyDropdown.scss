.currency-dropdown {
  &__button {
    font-size: 38px;
    margin-right: 20px;
    background-image: url(/img/arrow-down-icon.png);
    background-position: right center;
    background-repeat: no-repeat;
    background-size: 15px;
    padding-right: 20px;
    cursor: pointer;
    user-select: none;
  }

  &__list {
    position: absolute;
    background-color: white;
    box-shadow: 0 10px 20px 0 rgba(62, 131, 255, 0.15);
    border-radius: 10px;
    padding: 15px 30px;
    z-index: 1;
    margin: 0;
    visibility: hidden;
    transition: transform 200ms cubic-bezier(0.33, 1, 0.68, 1),
      opacity 300ms cubic-bezier(0.33, 1, 0.68, 1);
    transform: translateY(-30px);
    opacity: 0;

    &--shown {
      visibility: visible;
      opacity: 1;
      transform: translateY(8px);
    }
  }

  &__item {
    padding: 10px 0;
    font-size: 19px;
    cursor: pointer;
    list-style: none;

    &--active,
    &:hover {
      color: #0075eb;
    }
  }
}
