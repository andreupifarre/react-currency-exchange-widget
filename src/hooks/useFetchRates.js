import { useState, useEffect } from 'react';

const useFetchRates = (fromCurrency, interval) => {
  const [rates, setRates] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      const url = `https://openexchangerates.org/api/latest.json?app_id=81fa12e332b74f45a1de4a276db7c0d1&base=${fromCurrency}`;
      const response = await fetch(url);
      const data = await response.json();
      setRates(data.rates);
    };

    fetchData();
  }, [fromCurrency, interval]);

  return rates;
};

export default useFetchRates;
