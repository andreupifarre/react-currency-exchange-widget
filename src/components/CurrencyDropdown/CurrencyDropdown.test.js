import React from 'react';
import { shallow } from 'enzyme';

import CurrencyDropdown from './CurrencyDropdown';
// import useDebounce from '../../hooks/useDebounce';
import useFetchRates from '../../hooks/useFetchRates';

// jest.mock('../../hooks/useDebounce');
jest.mock('../../hooks/useFetchRates');

// useDebounce.mockReturnValue('');
useFetchRates.mockReturnValue([]);

const render = (props) => shallow(<CurrencyDropdown {...props} />);

afterEach(() => {
  jest.clearAllMocks();
});

describe('CurrencyDropdown', () => {
  it('should contain a parent element with a class .currency-dropdown', () => {
    const element = render().find('.currency-dropdown');
    expect(element.exists()).toBeTruthy();
  });

  it('should contain a span element with a class .currency-dropdown__button', () => {
    const element = render().find('.currency-dropdown__button');
    expect(element.exists()).toBeTruthy();
    expect(element.name()).toEqual('span');
  });

  it('should contain an ul element with a class .currency-dropdown__list', () => {
    const element = render().find('.currency-dropdown__list');
    expect(element.exists()).toBeTruthy();
    expect(element.name()).toEqual('ul');
  });

  it('should display list of items corresponding to the currencies passed', () => {
    const currencies = [
      { code: 'GBP', balance: '10', sign: '£' },
      { code: 'EUR', balance: '0', sign: '€' },
      { code: 'USD', balance: '0', sign: '$' },
    ];
    const currency = 'EUR';

    const wrapper = render({ currencies, currency });
    const button = wrapper.find('.currency-dropdown__button');

    button.simulate('click');

    const element = wrapper.find('.currency-dropdown__list');
    expect(element.find('.currency-dropdown__item').length).toEqual(3);
  });

  it('should display the label corresponding to the currency chosen', () => {
    const currencies = [
      { code: 'GBP', balance: '10', sign: '£' },
      { code: 'EUR', balance: '0', sign: '€' },
      { code: 'USD', balance: '0', sign: '$' },
    ];
    const currency = 'EUR';

    const wrapper = render({ currencies, currency });
    const button = wrapper.find('.currency-dropdown__button');

    button.simulate('click');

    const element = wrapper.find('.currency-dropdown__list');
    const secondItemElement = element.find('.currency-dropdown__item').at(1);
    secondItemElement.simulate('click');

    expect(button.text()).toEqual(currency);
  });

  it('should render correctly', () => {
    expect(render().getElements()).toMatchSnapshot();
  });
});
