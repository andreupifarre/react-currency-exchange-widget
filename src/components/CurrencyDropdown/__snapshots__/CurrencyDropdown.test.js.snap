// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyDropdown should render correctly 1`] = `
Array [
  <div
    className="currency-dropdown"
  >
    <span
      className="currency-dropdown__button"
      onClick={[Function]}
    >
      
    </span>
    <ul
      className="currency-dropdown__list"
    />
  </div>,
]
`;
